### ISSUE TITLE
Error while retrieving feed: no RSS version

### ISSUE THREAD

**Issue by tcpaiva** [newsboat/newsboat#1435]
**Newsboat version (copy the output of `newsboat -v` or the first line of `git show`)**:

```
$ newsboat -v
newsboat 2.18.0 - https://newsboat.org/
Copyright (C) 2006-2015 Andreas Krennmair
Copyright (C) 2015-2019 Alexander Batischev
Copyright (C) 2006-2017 Newsbeuter contributors
Copyright (C) 2017-2019 Newsboat contributors

Newsboat is free software licensed under the MIT License. (Type `newsboat -vv' to see the full text.)
It bundles JSON for Modern C++ library, licensed under the MIT License: https://github.com/nlohmann/json

newsboat 2.18.0
System: Linux 5.4.0-60-generic (x86_64)
Compiler: g++ 9.3.0
ncurses: ncurses 6.2.20200212 (compiled with 6.2)
libcurl: libcurl/7.68.0 GnuTLS/3.6.13 zlib/1.2.11 brotli/1.0.7 libidn2/2.2.0 libpsl/0.21.0 (+libidn2/2.2.0) libssh/0.9.3/openssl/zlib nghttp2/1.40.0 librtmp/2.3 (compiled with 7.68.0)
SQLite: 3.31.1 (compiled with 3.31.1)
libxml2: compiled with 2.9.10
```

Config file (copy from ~/.newsboat/config or ~/.config/newsboat/config):

```
browser "w3m -s -W -cols 80 -dump %u | less"
show-read-feeds  yes
auto-reload no
text-width 72
download-full-page yes
# pager "less %f"
max-items 25
```

Steps to reproduce the issue:

1. Include the following item in `urls` file: `http://rss.uol.com.br/feed/noticias.xml`

2. Open `newsboat` and ask to reload feeds

3. `newsboat` returns: `Error while retrieving http://rss.uol.com.br/feed/noticias.xml: no RSS version`

**Other info you think is relevant**:

The mentioned RSS feed can be read through other readers. Maybe `newsbloat` is too strict? Is there a way to overcome this situation? Thanks for your great work!

**Comment by Minoru** [comment]
Thank you for the report! Reproduced with 2.22.1.

The feed is indeed missing a `version` attribute in the `<rss>` tag. There are at least 5 slightly different versions of RSS, so it's not clear what default to pick. Are any of those other readers that you mentioned open-source? Perhaps I can peek at their code and see what they do :)

Furthermore, the feed uses ISO-8859-1 encoding and signals that through the `Content-Type` header, but Newsboat doesn't actually look into that. I think this is a violation of XML spec, and I'll open a separate issue about that.

Anyway, to dodge all the bugs, write a script _scriptname.sh_:

```
#!/bin/sh

echo '<?xml version="1.0" encoding="UTF-8" ?>'
curl --silent "http://rss.uol.com.br/feed/noticias.xml" | iconv -f iso-8859-1 -t utf-8 | sed 's#<rss>#<rss version="2.0">#'
```

(Version 2.0 seems to work, but I didn't look too deep to see if the feed uses any non-2.0 elements.)

Then add something like this to your urls file:

```
filter:~/scriptname.sh
```

**Comment by dennisschagt** [comment]
@Minoru: Regarding `filter:~/scriptname.sh`, I believe this should contain `exec:~/scriptname.sh` instead.
A `filter:` line requires an RSS URL. In that case, Newsboat downloads the feed and provides it to the script on stdin.
On the other hand, your example script downloads the feed independently so an `exec:` line should be enough.

Relevant part of the documentation: https://newsboat.org/releases/2.22.1/docs/newsboat.html#_scripts_and_filters_snownews_extensions

**Comment by Minoru** [comment]
Right, thanks! JFTR, we can't use `filter:` here because Newsboat thinks that the input to the filter is UTF-8, and panics when it isn't so. That'll be reported too, I'm just busy figuring out the easiest way to mock an HTTP server that returns a `Content-Type` header that I want.

**Comment by der-lyse** [comment]
@Minoru, an easy mock could be made with netcat, something along those lines on Debian (please note the `nc` implementations vary on different systems):

```bash
(
content_length=$(wc -c feed.rss | cut -d ' ' -f 1)
printf "HTTP/1.1 200 OK\r\nContent-Type: application/json; charset=my-super-charset\r\nContent-Length: $content_length\r\n\r\n"
cat feed.rss
) | nc -l 9999
```

Works nicely with curl:
```bash
$ curl http://localhost:9999/ -D -
HTTP/1.1 200 OK
Content-Type: application/json; charset=my-super-charset
Content-Length: 47

…contents from feeds.rss…
```

The only drawback is, that the server terminates after sending the response, so a restart loop may be needed.

**Comment by Minoru** [comment]
@der-lyse, yeah, I did a similar thing :) But thanks for point it out — I might've not come up with it myself.

**Comment by tcpaiva** [comment]
Thanks for looking into this issue and for your prompt answer.

The other readers I know work with this are `newsblur` and `akregator`.

One question, though, how do I use these suggested scripts with newsboat? Can they be used with the other feeds that are listed in the `urls` file?

**Comment by Minoru** [comment]
> The other readers I know work with this are newsblur and akregator.

Ah, cool! Newsblur [is open source](https://github.com/samuelclay/NewsBlur), and Akgregator seem to be a part of KDE so its code should be somewhere too.

> One question, though, how do I use these suggested scripts with newsboat? Can they be used with the other feeds that are listed in the urls file?

You put the script wherever you like, make it executable with `chmod +x script_name.sh`, and mention the script in your urls file as I described. Every time you update the feed, Newsboat will execute the script and treat its output as a feed.

To use it with other feeds, you'll have to modify the script somehow. I wrote it specifically for the problematic feed, so everything is hardcoded. If you have more feeds with the *exact* same problem, you can turn the URL into a command-line parameter and change the urls file like this:

```
"exec:~/scriptname.sh https://example.com/feed.xml"
```

I see you have some Python repos — scripts can be written in Python too! They just need to print out the feed to stdout, that's all that's required of them.

See the docs for a bit more explanation of this: https://newsboat.org/releases/2.22.1/docs/newsboat.html#_scripts_and_filters_snownews_extensions

**Comment by tcpaiva** [comment]
Good, thanks! I didn't read your instructions carefully enough.  The provided solution worked nicely.
